### PR TITLE
[WIP] Add Folder based authorization

### PR DESCRIPTION
**Pull request by AbhyudayaSharma**
Add ability to organize permissions at the folder level. The plan is to provide permissions through folders Instead of using regular expressions to match projects.

**This is a work-in-progress pull request**

### CODE REVIEW

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RegexAuthorizationEngine.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package com.michelin.cio.hudson.plugins.rolestrategy;
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RegexAuthorizationEngine.java]: I would recommend keeping new code in one of the new packages

Comment (AbhyudayaSharma) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RegexAuthorizationEngine.java]: How about `io.jenkins.plugins.rolestrategy`? Or should I stick with `org.jenkinsci.plugins.rolestrategy`?

**Comment by AbhyudayaSharma** [comment]
I'm sorry about the large changes to the Jelly files. Git's `autocrlf true` changed CRLF to LF when I committed. If you like, I can revert them.

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 2]
```diff
@@ -644,16 +669,13 @@ public static RoleBasedAuthorizationStrategy getInstance() {
      * Updates macro roles
      * @since 2.1.0
      */
-    void renewMacroRoles()
-    {
+    void renewMacroRoles() {
```
Comment (AbhyudayaSharma) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java]: Is this function needed any more? It doesn't seem to do anything.

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java [hunk 3]
```diff
@@ -474,34 +478,55 @@ public void doDeleteSid(@QueryParameter(required = true) String type,
...
+                throw new IllegalArgumentException("Unsupported RoleType");
+        }
+
+        persistChanges();
     }
 
     /**
-     * API method to get all groups/users with their role in any role type
-     * Example: curl -X GET localhost:8080/role-strategy/strategy/getAllRoles?type=projectRoles
+     * Gets the sids assigned to each role.
```
Comment (AbhyudayaSharma) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java]: I've changed the JavaDoc, it did not seem correct.

File: src/main/java/io/jenkins/plugins/rolestrategy/AbstractRole.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package io.jenkins.plugins.rolestrategy;
+
+import hudson.security.Permission;
+
+import java.util.Set;
+
+public abstract class AbstractRole {
```
Comment (AbhyudayaSharma) [inline src/main/java/io/jenkins/plugins/rolestrategy/AbstractRole.java]: This class may not be needed in the future and is here as a placeholder for the time being.

**Comment by AbhyudayaSharma** [comment]
https://github.com/jenkinsci/role-strategy-plugin/blob/9b4e1bab52e68b0ca2cd865b70625c4b615b3608/src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleBasedAuthorizationStrategy.java#L91-L92

Can these be removed (or marked deprecated)? I can find no references to it.

File: src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/assign-roles.jelly [hunk 5]
```diff
@@ -1,100 +1,103 @@
...
-            </f:section>
-            <f:block>
-              <f:submit value="${%Save}" />
-              <f:apply />
-            </f:block>
-          </f:form>
-        </l:main-panel>
-    </l:layout>
-</j:jelly>
+<!--
```
Comment (oleg-nenashev) [inline src/main/resources/com/michelin/cio/hudson/plugins/rolestrategy/RoleStrategyConfig/assign-roles.jelly]: hm? Formatting change?

**Comment by AbhyudayaSharma** [comment]
This pull request was superseded by #89 and the code for the new plugin was moved to https://github.com/jenkinsci/folder-auth-plugin . It can be reopened when we need to make Role Strategy Plugin more extensible.

**Pull request closed without merging**